README
kettle-cookbook
===============
Kettle-Cookbook is an auto-documentation tool for Kettle, a.k.a. 
Pentaho Data Integration (http://kettle.pentaho.com/).
It generates human readable documentation in HTML format for 
collections of Kettle .ktr / .kjb files. 

Kettle-cookbook is open source (LGPL), and is available from 
google code: http://code.google.com/p/kettle-cookbook/

Kettle-cookbook is itself built as a Kettle solution. 
To use kettle-cookbook, you have to run the document-all.kjb kettle
job located in the pdi directory. The job accepts two parameters:

INPUT_DIR: the directory containing the .ktr/.kjb files
OUTPUT_DIR: the directory where the generated documentation will be 
stored.

You can run the document-all.kjb from the kettle IDE spoon, or run
it from the command line using the kitchen tool. For more detailed
instructions, see the wiki pages on the google code project:

http://code.google.com/p/kettle-cookbook/wiki/CookBookInstructions

Roland Bouman
blog: http://rpbouman.blogspot.com/
email: roland.bouman@gmail.com
twitter: @rolandbouman

change_log.txt
This file lists changes done ontop of r79

xslt/Kettle_Report.xslt
   - added imported io_steps.xslt
   - unconditionally display connections (<xsl:if test="$usages"> removed)
   - fixed some bugs
       - removed some un-used code
       - some incorrect XPaths
       - connections now reported as links
       - steps in highlevel flow diagram now reported as links
       - info section for transformations was missing
       - Highlevel flow diagram now contains links to steps
       - outcommented flat files (not yet supported)
   - improved reporting of steps:
       - steps separated with horizontal lines
       - steps now report hops (to/from) in a separate sub-section
       - mappings (ETL Metadata Injection), lookups, database tables
         (this solves most of issue 37), all tabs of 'Select Values' step
       
xslt/files.xslt
   - define variable 'modified_date' based on file property 

xslt/shared.xslt
   - template description changed 
     - maintain formatting with <pre>
     - different font if 'no description'

xslt/toc.xslt
   - outcommented yet unsupported sections

document-folder.kjb
   - none

process-files.ktr
   - added (but disabled) flow for full re-build (instead of 'make')

template/index.html
   - space reserved for frame on left adapted

template/default.css
   - a:hover constrained to [href]

template/kettle.css
   - changed font for .description, .extended-description to courier (blue)

template/kettle.css, job.css
   - created with constructor job

template/images
   - updated to kettle 4.4.0
   
constructor/*
   - used to adapt kettle-cookbook to new versions of kettle
     (see job descritption of Create_Files.kjb for details)
   - contains sub-dir ./AllOperators with a bunch of jobs and transformations
     simply instantiating all operators - helps visualizing what the constructor
     misses (namely, plug-ins)

template/js/kettle.js
/**
*   This is kettle.js. This is part of kettle-cookbook.
*   Kettle-cookbook is distributed on http://code.google.com/p/kettle-cookbook/
*   
*   kettle.js - a javascript file that helps draw the hops for the diagrams
*   
*   Copyright (C) 2010 
*   Roland Bouman Roland.Bouman@gmail.com - http://rpbouman.blogspot.com/
*   Chodnicki Slawomir (arrows in hops)
*   David Bouyssel, Samatar Hassan (IE8 / IE7 compatibility)
*   
*   This library is free software; you can redistribute it and/or modify it under 
*   the terms of the GNU Lesser General Public License as published by the 
*   Free Software Foundation; either version 2.1 of the License, or (at your option)
*   any later version.

*   This library is distributed in the hope that it will be useful, but WITHOUT ANY 
*   WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A 
*   PARTICULAR PURPOSE. See the GNU Lesser General Public License for more details.
*   
*   You should have received a copy of the GNU Lesser General Public License along 
*   with this library; if not, write to 
*   the Free Software Foundation, Inc., 
*   59 Temple Place, Suite 330, 
*   Boston, MA 02111-1307 USA
*   
*/
var regexpStepHopCopyData = /\bstep-hop-copy-data\b/,               //used for steps that copy data
    regexpStepHopDistributeData = /\bstep-hop-distribute-data\b/,   //used for steps that distribute data
    regexpStepHopTrue = /\b(step|entry)-hop-true\b/,         //used for hops that route "true" data
    regexpStepHopFalse = /\b(step|entry)-hop-false\b/,              //used for hops that route "false" data
    regexpEntryHopUnconditional = /\bentry-hop-unconditional\b/     //used for unconditional job entry hops.
;

function alignTowardsLine(array_x, array_y, origin_x, origin_y, line_x1, line_y1, line_x2, line_y2){
	
	// line vector
	var vx = line_x2-line_x1;
	var vy = line_y2-line_y1;
	
	// normalize it
	var vlen = Math.sqrt(vx*vx+vy*vy);
	vx /= vlen;
	vy /= vlen;
	
	// determine angle
	var angle = Math.acos(vy);
	if (vx > 0){
		angle = -angle;
	}
	
	// apply same angle to all points
	var res_x = [];
	var res_y = [];
	
	for(var i=0;i<array_x.length;i++){
		// center of object
		var cx = array_x[i] - origin_x;
		var cy = array_y[i] - origin_y;
		
		// rotate and offset to position
		res_x[i] = cx * Math.cos(angle) - cy * Math.sin(angle) + origin_x;
		res_y[i] = cx * Math.sin(angle) + cy * Math.cos(angle) + origin_y;
		
	}

	return {
		array_x: res_x,
		array_y: res_y
	};
	
}

// function log(txt){
// 	var d = document.getElementById("logdiv");
// 	d.innerHTML += txt+"<br/>"
// }

function drawHops(){
	
    var idCanvas = "canvas",
		id = "thediagram",
        jsg = new jsGraphics(idCanvas),
        diagram = document.getElementById(id),
        childNodes = diagram.childNodes,
        numChildNodes = childNodes.length,
        from,
        i,
        hops,
        j,
        numHops,
        hop,
        to,
		offset = 16,
		x1, y1, x2, y2,
		xDiff, yDiff,
		m = Math, pi2 = 2 * m.PI,
		angle,
		arrowWidth = 10,
		arrowHeight = 20,
		arrowPos = 0.59,
        className, copyData
    ;
    jsg.setClassNames("kettle-hop");
    for (i = 0; i<numChildNodes; i++) {
        from = childNodes.item(i);
        if (from.nodeType !== 1         //element node
        ||  from.tagName  !== "DIV"
        ||  from.className==="note"
        ) {
            continue;
        }
		x1 = parseInt(from.style.left, 10) + offset;
		y1 = parseInt(from.style.top, 10)  + offset;

        hops = from.getElementsByTagName("DIV").item(0);
        hops = hops.getElementsByTagName("A");
		
		numHops = hops.length;

        for (j = 0; j < numHops; j++){
            hop = hops.item(j);
            
            to = hop.getAttribute("href");
			to = to.substring(to.lastIndexOf("#"));
            to = decodeURIComponent(to.substring(1));
            to = document.getElementById(to);

			x2 = parseInt(to.style.left, 10) + offset;
            xDiff = x2 - x1;
			y2 = parseInt(to.style.top, 10) + offset;
            yDiff = y2 - y1;

            className = hop.className;

            jsg.setClassNames(className);
            jsg.drawLine(x1, y1, x2, y2);
			
			// calculate placement the way from source to target
			var xMid = x1 + arrowPos * xDiff;
			var yMid = y1 + arrowPos * yDiff;
			
			// place a triangle there
			var xLeft = xMid - arrowWidth/2;
			var yLeft = yMid;
			
			var xRight = xMid + arrowWidth/2;
			var yRight = yMid;
			
			var xTop = xMid;
			var yTop = yMid + arrowHeight;
			
			var polygon_x = [xLeft, xTop, xRight];
			var polygon_y = [yLeft, yTop, yRight];
			
			var aligned = alignTowardsLine(polygon_x, polygon_y, xMid, yMid, x1, y1, x2, y2);
			jsg.fillPolygon(aligned.array_x, aligned.array_y);
			    
            if (regexpStepHopTrue.test(className) ) {
                hopIcon = "hop-true-icon";
            }
            else
            if (regexpStepHopFalse.test(className) ) {
                hopIcon = "hop-false-icon";
            }
            else
            if (regexpEntryHopUnconditional.test(className) ) {
                hopIcon = "entry-hop-unconditional-icon";
            }
            else
            if (regexpStepHopCopyData.test(className) ) {
                hopIcon = "step-hop-copy-data-icon";
            }
            else {
                hopIcon = false;
            }
            
            if (hopIcon) {
                hopIconEl = document.createElement("DIV");
                hopIconEl.className = hopIcon;
                hopIconEl.style.left = x1 + (xDiff * .33) - 8 + "px";
                hopIconEl.style.top = y1 + (yDiff * .33) - 8 + "px";
                diagram.appendChild(hopIconEl);
            }
        }
    }
    jsg.paint();

}


template/js/kettle-toc.js
/**
*   This is kettle-toc.js. This is part of kettle-cookbook.
*   Kettle-cookbook is distributed on http://code.google.com/p/kettle-cookbook/
*   
*   kettle-toc.js - javascript to implement tab page and treeview interactivity
*   for the table of contents (toc)
*   
*   Copyright (C) 2010 
*   Roland Bouman Roland.Bouman@gmail.com - http://rpbouman.blogspot.com/
*   
*   This library is free software; you can redistribute it and/or modify it under 
*   the terms of the GNU Lesser General Public License as published by the 
*   Free Software Foundation; either version 2.1 of the License, or (at your option)
*   any later version.

*   This library is distributed in the hope that it will be useful, but WITHOUT ANY 
*   WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A 
*   PARTICULAR PURPOSE. See the GNU Lesser General Public License for more details.
*   
*   You should have received a copy of the GNU Lesser General Public License along 
*   with this library; if not, write to 
*   the Free Software Foundation, Inc., 
*   59 Temple Place, Suite 330, 
*   Boston, MA 02111-1307 USA
*   
*/

function tabClicked(tabLink) {
    var pageId = tabLink.getAttribute("href").substr(2) + "-page",
        tab = tabLink.parentNode,
        tabs = document.getElementById("tab-strip").childNodes,
        pages = document.getElementById("tab-pages").childNodes,
        numPages, el, pageIndex, display
    ;
    
    for (pageIndex = 0, numPages = tabs.length; pageIndex < numPages; pageIndex++){
        el = tabs.item(pageIndex);
        if (el.nodeType!==1 || el.className==="tab-spacer") {
            continue;
        }
        el.className = ((el===tab) ? "tab tab-active" : "tab");
    }

    for (pageIndex = 0, numPages = pages.length; pageIndex < numPages; pageIndex++){
        el = pages.item(pageIndex);
        if (el.nodeType!==1) {
            continue;
        }
        el.style.display = ((el.id === pageId) ? "" : "none");
    }
}

function toggleTreeNode(toggle) {

}

template/js/shBrushMdx.js
/**
 * SyntaxHighlighter
 * http://alexgorbatchev.com/SyntaxHighlighter
 *
 * SyntaxHighlighter is donationware. If you are using it, please donate.
 * http://alexgorbatchev.com/SyntaxHighlighter/donate.html
 *
 * @version
 * 3.0.83 (July 02 2010)
 * 
 * @copyright
 * Copyright (C) 2004-2010 Alex Gorbatchev.
 *
 * @license
 * Dual licensed under the MIT and GPL licenses.
 */
;(function()
{
	// CommonJS
	typeof(require) != 'undefined' ? SyntaxHighlighter = require('shCore').SyntaxHighlighter : null;

	function Brush()
	{
		var keywords =	'ABSOLUTE ACTIONPARAMETERSET AFTER ALL AND AS ASC AVERAGE BASC BDESC ' + 
                        'BEFORE BEFORE_AND_AFTER BY CACHE CALCULATE CALCULATION CALCULATIONS ' +
                        'CALL CELL CELLFORMULASETLIST CHAPTERS CLEAR COALESCEEMPTY COLUMN COLUMNS '+
                        'CREATE CREATEPROPERTYSET CREATEVIRTUALDIMENSION CUBE CURRENTCUBE ' +
                        'DEFAULT_MEMBER DESC DESCRIPTION DROP EMPTY END EXCLUDEEMPTY FALSE '+
                        'FOR FREEZE FROM GLOBAL GROUP GROUPING HIDDEN  IGNORE INCLUDEEMPTY INDEX ' +
                        'IS MEASURE MEMBER NEST NO_ALLOCATION NO_PROPERTIES NON NOT_RELATED_TO_FACTS '+
                        'NULL ON OR PAGES PASS POST PROPERTY QTD RECURSIVE RELATIVE ROWS SCOPE '+
                        'SECTIONS SELECT SELF SELF_AND_AFTER SELF_AND_BEFORE SELF_BEFORE_AFTER '+
                        'SESSION SET SETTOARRAY SORT STORAGE STRTOVAL TOTALS TREE TRUE TYPE '+
                        'UNIQUE UPDATE USE USE_EQUAL_ALLOCATION USE_WEIGHTED_ALLOCATION '+
                        'USE_WEIGHTED_INCREMENT VISUAL WHERE WITH XOR'
						;
		var functions =	'AddCalculatedMembers Aggregate AllMembers Ancestor Ancestors Ascendants '+
                        'Avg Axis BottomCount BottomPercent BottomSum CalculationCurrentPass '+
                        'CalculationPassValue Children ClosingPeriod '+
                        'CoalesceEmpty Correlation Count '+
                        'Cousin Covariance CovarianceN Crossjoin Current CurrentMember CurrentOrdinal '+
                        'DataMember DefaultMember Descendants Dimension Dimensions Distinct DistinctCount '+
                        'DrilldownLevel DrilldownLevelBottom DrilldownLevelTop DrilldownMember '+
                        'DrilldownMemberBottom DrilldownMemberTop DrillupLevel DrillupMember Error '+
                        'Except Exists Extract Filter FirstChild FirstSibling Generate Head '
                        'Hierarchize Hierarchy IIf Intersect IsAncestor IsEmpty '+
                        'IsGeneration IsLeaf IsSibling Item Lag LastChild LastPeriods LastSibling '+
                        'Lead Leaves Level Levels LinkMember LinRegIntercept LinRegPoint '+
                        'LinRegR2 LinRegSlope LinRegVariance LookupCube Max Median Members '+
                        'MemberToStr Min Mtd Name NameToSet NextMember NonEmptyCrossjoin '+
                        'OpeningPeriod Order Ordinal ParallelPeriod Parent PeriodsToDate Predict PrevMember '+
                        'Properties Qtd Rank RollupChildren Root SetToArray SetToStr Siblings Stddev StddevP '+
                        'Stdev StdevP StripCalculatedMembers StrToMember StrToSet StrToTuple StrToValue '+
                        'Subset Sum Tail This ToggleDrillState TopCount TopPercent TopSum TupleToStr '+
                        'Union UniqueName UnknownMember Unorder UserName ValidMeasure Value Var Variance '+
                        'VarianceP VarP VisualTotals Wtd Ytd '
						;

		var r = SyntaxHighlighter.regexLib;
		
		this.regexList = [
			{ regex: r.multiLineDoubleQuotedString,					css: 'string' },			// double quoted strings
			{ regex: r.multiLineSingleQuotedString,					css: 'string' },			// single quoted strings
			{ regex: r.singleLineCComments,							css: 'comments' },			// one line comments
			{ regex: r.multiLineCComments,							css: 'comments' },			// multiline comments
			{ regex: new RegExp(this.getKeywords(keywords), 'gmi'),	css: 'keyword' },			// keywords
			{ regex: new RegExp(this.getKeywords(functions), 'gmi'),css: 'color1' } 		    // functions
			];
	
		this.forHtmlScript(r.scriptScriptTags);
	};

	Brush.prototype	= new SyntaxHighlighter.Highlighter();
	Brush.aliases	= ['mdx'];

	SyntaxHighlighter.brushes.Mdx = Brush;

	// CommonJS
	typeof(exports) != 'undefined' ? exports.Brush = Brush : null;
})();


template/js/toc.js
/**
*   This is kettle-toc.js. This is part of kettle-cookbook.
*   Kettle-cookbook is distributed on http://code.google.com/p/kettle-cookbook/
*   
*   kettle-toc.js - javascript to implement tab page and treeview interactivity
*   for the table of contents (toc)
*   
*   Copyright (C) 2010 
*   Roland Bouman Roland.Bouman@gmail.com - http://rpbouman.blogspot.com/
*   
*   This library is free software; you can redistribute it and/or modify it under 
*   the terms of the GNU Lesser General Public License as published by the 
*   Free Software Foundation; either version 2.1 of the License, or (at your option)
*   any later version.

*   This library is distributed in the hope that it will be useful, but WITHOUT ANY 
*   WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A 
*   PARTICULAR PURPOSE. See the GNU Lesser General Public License for more details.
*   
*   You should have received a copy of the GNU Lesser General Public License along 
*   with this library; if not, write to 
*   the Free Software Foundation, Inc., 
*   59 Temple Place, Suite 330, 
*   Boston, MA 02111-1307 USA
*   
*/

function tabClicked(tabLink) {
    var pageId = tabLink.getAttribute("href").substr(2) + "-page",
        tab = tabLink.parentNode,
        tabs = document.getElementById("tab-strip").childNodes,
        pages = document.getElementById("tab-pages").childNodes,
        numPages, el, pageIndex, display
    ;
    
    for (pageIndex = 0, numPages = tabs.length; pageIndex < numPages; pageIndex++){
        el = tabs.item(pageIndex);
        if (el.nodeType!==1 || el.className==="tab-spacer") {
            continue;
        }
        el.className = ((el===tab) ? "tab tab-active" : "tab");
    }

    for (pageIndex = 0, numPages = pages.length; pageIndex < numPages; pageIndex++){
        el = pages.item(pageIndex);
        if (el.nodeType!==1) {
            continue;
        }
        el.style.display = ((el.id === pageId) ? "" : "none");
    }
}

function toggleTreeNode(toggle) {
    var els, numEls, el, elIndex,
        oldState, newState, oldIconClass, newIconClass, display
    ;
    switch (oldState = toggle.innerHTML){
        case "+":
            oldIconClass = "folder-icon-closed";
            newIconClass = "folder-icon-open";
            newState = "-";
            display = "";
            break;
        case "-":
            oldIconClass = "folder-icon-open";
            newIconClass = "folder-icon-closed";
            newState = "+";
            display = "none";
            break;
    }
    
    els = toggle.parentNode.childNodes, numEls = els.length;
    for (elIndex = 0; elIndex < numEls; elIndex++){
        el = els.item(elIndex);
        if  (el.nodeType!==1 || el.className!==oldIconClass) {
            continue;
        }
        el.className = newIconClass;
        break;
    }
    
    els = toggle.parentNode.parentNode.childNodes, numEls = els.length;
    for (elIndex = 0; elIndex < numEls; elIndex++){
        el = els.item(elIndex);
        if (el.nodeType!==1 || el.className!=="folder-body"){
            continue;
        }
        el.style.display = display;
        break;
    }
    toggle.innerHTML = newState;
}